subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DefaultFileSystemSnapshotter.java

```diff
@@ -232,14 +231,15 @@ private FileSystemSnapshot filterSnapshot(FileSystemLocationSnapshot snapshot, P
         return FileSystemSnapshotFilter.filterSnapshot(spec, snapshot, fileSystem);
     }
 
-    private class FileCollectionLeafVisitorImpl implements FileCollectionLeafVisitor {
-        private final List<FileSystemSnapshot> roots = new ArrayList<FileSystemSnapshot>();
+    private class RootVisitingSnapshotBuilder implements FileCollectionLeafVisitor {
+        private final ImmutableSet.Builder<FileSystemSnapshot> roots = ImmutableSet.builder();
```

Comment:
My understanding is that `snapshot()` should return the same instance for the same path every time, so the `Set` will not contain duplicates. I opted for a `Set` here because we don't de-dupe via `FileCollection.getFiles()` anymore, so I wanted to de-dupe here "globally" instead.

Do you think we need to keep track of the root paths instead, and make sure we only `snapshot()` each once?

Reply:
I think that would be better. Moreover, having the same root multiple times is currently an option, isn't it? You could register the same input directory twice. I think it is fine if you only have it once in the end when we de-dupe on snapshotting here.

I think the main problem is that the file system mirror can be invalidated while we snapshot a file collection when another task in another project starts executing. That means that `snapshot` will return a new instance of the file system snapshot.

Reply:
IIUC, if the `FSM` is invalidated in-between snapshotting the same root twice, with the old code we'd end up with a duplicate snapshot, too (if the same root comes from different sub-filecollections, so `getFiles()` doesn't have a chance to de-dupe them away). It's broken both ways, basically. We should think about how to address that use case separately.

> I think that would be better.

I think I disagree here. De-duping the paths up front would only make a difference if `FSM` was invalidated while we are snapshotting, but that's already problematic. We should fix that separately.

> Moreover, having the same root multiple times is currently an option, isn't it? You could register the same input directory twice. I think it is fine if you only have it once in the end when we de-dupe on snapshotting here.

How is having the same root multiple times an option now? The only access the user has to the `FileCollection` that we snapshot and fingerprint will de-dupe. So registering the same directory twice doesn't get reflected, as they will only see it once. The fact that our snapshotting/fingerprinting currently doesn't de-dupe this is a bug that is fixed by this change.

Reply:
For me, the important thing is that we always have an equal List/Set of file system snapshots for the same file collection, no matter what else happens in the build. The current code may produce a different number of roots depending on whether the file system mirror is invalidated in between or not.

Producing the same set of roots can be achieved by
- Dedupe per absolute path (`File` object) in `visitFileCollection`
- Add an equals method to `FileSystemSnapshot` and de-dupe globally
- Do not de-dupe at all
- Make sure the file system mirror is not invalidated while snapshotting a leaf file collection

I would choose the first option since it seems to be easy and it is the same what we have on `master` already.

When adding the same root via two different file collections to a composite file collection, the root is snapshotted twice and added twice as a root. So if you do something like this you have the root `duplicate` twice:
```
class MyTask extends DefaultTask {
    @InputFiles
    ConfigurableFileCollection inputFiles
    ...
}

myTask.inputFiles.from(files("duplicate"))
myTask.inputFiles.from(files("duplicate"))
```

I also don't think that invalidating the file system mirror while we are snapshotting one file collection is very rare, especially if you have a larger multiproject build. Given that we always invalidate all of the file system mirror when some task is about to execute, I think that is what currently happens.

> De-duping the paths up front would only make a difference if FSM was invalidated while we are snapshotting, but that's already problematic. We should fix that separately.

Is it problematic for performance reasons? It totally is. I don't see us fixing this in the near future, though.

> The fact that our snapshotting/fingerprinting currently doesn't de-dupe this is a bug that is fixed by this change.

Fingerprinting does de-dupe, I don't see how the snapshotter not de-duping is a bug.

Reply:
FTR, we agreed not to de-dupe in the snapshotter at all. Let's see if this leads to a regression.